lib/os/onoff.c

```diff
@@ -0,0 +1,542 @@
+
+	/* Can't have a queued releaser during start */
+	__ASSERT_NO_MSG(srv->releaser == NULL);
+
+	/* If the start failed log an error and leave the rest of the
+	 * state in place for diagnostics.
+	 *
+	 * If the start succeeded record a reference for all clients
+	 * and set the state to ON.  But if there are no clients,
+	 * initiate a transition to off.
+	 *
+	 * In either case reset the client queue and notify all
+	 * clients of operation completion.
+	 */
+	if (res < 0) {
+		srv->flags &= ~SERVICE_STATE_TRANSITION;
+		srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+	} else {
+		sys_snode_t *node;
+
+		SYS_SLIST_FOR_EACH_NODE(&clients, node) {
+			srv->refs += 1;
+		}
+		set_service_state(srv, SERVICE_STATE_ON);
+		stop = (srv->refs == 0);
+		if ((stop || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_STOP_WAITS) != 0U)) {
+			srv->flags |= ONOFF_SERVICE_HAS_ERROR;
+			stop = false;
+		}
+	}
+
+	sys_slist_init(&srv->clients);
+
+	k_spin_unlock(&srv->lock, key);
+
+	notify_all(srv, &clients, res);
+	if (stop) {
+		srv->stop(srv, onoff_stop_notify);
+	}
+}
+
+int z_impl_onoff_request(struct onoff_service *srv,
+			 struct onoff_client *cli)
+{
+	bool add_client = false;        /* add client to pending list */
+	bool start = false;             /* invoke start transition */
+	bool notify = false;            /* do client notification */
+	bool no_wait = false;
+	int rv = validate_args(srv, cli, &no_wait);
+
+	if (rv != 0) {
+		return rv;
+	}
+
+	k_spinlock_key_t key = k_spin_lock(&srv->lock);
+
+	if ((srv->flags & ONOFF_SERVICE_HAS_ERROR) != 0) {
+		rv = -EIO;
+		goto out;
+	}
+
+	if (srv->refs == UINT16_MAX) {
+		rv = -EAGAIN;
+		goto out;
+	}
+
+	u32_t state = srv->flags & SERVICE_STATE_MASK;
+
+	switch (state) {
+	case SERVICE_STATE_TO_OFF:
+		/* Reject if no-wait because operation will not be
+		 * completed when this call returns).
+		 */
+		if (no_wait) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Queue to start after release */
+		__ASSERT_NO_MSG(srv->releaser != NULL);
+		add_client = true;
+		rv = 3;
+		break;
+	case SERVICE_STATE_OFF:
+		/* Reject if in a no-wait context and start could
+		 * wait.
+		 */
+		if ((no_wait || k_is_in_isr())
+		    && ((srv->flags & ONOFF_SERVICE_START_WAITS) != 0U)) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Start with first request while off */
+		__ASSERT_NO_MSG(srv->refs == 0);
+		set_service_state(srv, SERVICE_STATE_TO_ON);
+		start = true;
+		add_client = true;
+		rv = 2;
+		break;
+	case SERVICE_STATE_TO_ON:
+		/* Reject if no-wait because operation will not be
+		 * completed when this call returns.
+		 */
+		if (no_wait) {
+			rv = -EWOULDBLOCK;
+			break;
+		}
+
+		/* Already starting, just queue it */
+		add_client = true;
+		rv = 1;
+		break;
+	case SERVICE_STATE_ON:
+		/* Just increment the reference count */
+		notify = true;
+		break;
+	default:
+		rv = -EINVAL;
+		break;
+	}
+
+out:
+	if (add_client) {
+		sys_slist_append(&srv->clients, &cli->node);
+	} else if (notify) {
```

Comment:
`else if (notify || !no_wait)`?
In case of `no wait` client counter should also be incremented.

Reply:
That will be done in `onoff_notify_start()` because the no-wait client is in the list.